README.md
## Overview

Docker image packaging for oxShibboleth.

## Versions

- Stable: `gluufederation/oxshibboleth:4.2.0_02`
- Unstable: `gluufederation/oxshibboleth:4.2.0_dev`

Refer to [Changelog](./CHANGES.md) for details on new features, bug fixes, or older releases.

## Environment Variables

The following environment variables are supported by the container:

- `GLUU_CONFIG_ADAPTER`: The config backend adapter, can be `consul` (default) or `kubernetes`.
- `GLUU_CONFIG_CONSUL_HOST`: hostname or IP of Consul (default to `localhost`).
- `GLUU_CONFIG_CONSUL_PORT`: port of Consul (default to `8500`).
- `GLUU_CONFIG_CONSUL_CONSISTENCY`: Consul consistency mode (choose one of `default`, `consistent`, or `stale`). Default to `stale` mode.
- `GLUU_CONFIG_CONSUL_SCHEME`: supported Consul scheme (`http` or `https`).
- `GLUU_CONFIG_CONSUL_VERIFY`: whether to verify cert or not (default to `false`).
- `GLUU_CONFIG_CONSUL_CACERT_FILE`: path to Consul CA cert file (default to `/etc/certs/consul_ca.crt`). This file will be used if it exists and `GLUU_CONFIG_CONSUL_VERIFY` set to `true`.
- `GLUU_CONFIG_CONSUL_CERT_FILE`: path to Consul cert file (default to `/etc/certs/consul_client.crt`).
- `GLUU_CONFIG_CONSUL_KEY_FILE`: path to Consul key file (default to `/etc/certs/consul_client.key`).
- `GLUU_CONFIG_CONSUL_TOKEN_FILE`: path to file contains ACL token (default to `/etc/certs/consul_token`).
- `GLUU_CONFIG_KUBERNETES_NAMESPACE`: Kubernetes namespace (default to `default`).
- `GLUU_CONFIG_KUBERNETES_CONFIGMAP`: Kubernetes configmaps name (default to `gluu`).
- `GLUU_CONFIG_KUBERNETES_USE_KUBE_CONFIG`: Load credentials from `$HOME/.kube/config`, only useful for non-container environment (default to `false`).
- `GLUU_SECRET_ADAPTER`: The secrets adapter, can be `vault` or `kubernetes`.
- `GLUU_SECRET_VAULT_SCHEME`: supported Vault scheme (`http` or `https`).
- `GLUU_SECRET_VAULT_HOST`: hostname or IP of Vault (default to `localhost`).
- `GLUU_SECRET_VAULT_PORT`: port of Vault (default to `8200`).
- `GLUU_SECRET_VAULT_VERIFY`: whether to verify cert or not (default to `false`).
- `GLUU_SECRET_VAULT_ROLE_ID_FILE`: path to file contains Vault AppRole role ID (default to `/etc/certs/vault_role_id`).
- `GLUU_SECRET_VAULT_SECRET_ID_FILE`: path to file contains Vault AppRole secret ID (default to `/etc/certs/vault_secret_id`).
- `GLUU_SECRET_VAULT_CERT_FILE`: path to Vault cert file (default to `/etc/certs/vault_client.crt`).
- `GLUU_SECRET_VAULT_KEY_FILE`: path to Vault key file (default to `/etc/certs/vault_client.key`).
- `GLUU_SECRET_VAULT_CACERT_FILE`: path to Vault CA cert file (default to `/etc/certs/vault_ca.crt`). This file will be used if it exists and `GLUU_SECRET_VAULT_VERIFY` set to `true`.
- `GLUU_SECRET_KUBERNETES_NAMESPACE`: Kubernetes namespace (default to `default`).
- `GLUU_SECRET_KUBERNETES_CONFIGMAP`: Kubernetes secrets name (default to `gluu`).
- `GLUU_SECRET_KUBERNETES_USE_KUBE_CONFIG`: Load credentials from `$HOME/.kube/config`, only useful for non-container environment (default to `false`).
- `GLUU_WAIT_MAX_TIME`: How long the startup "health checks" should run (default to `300` seconds).
- `GLUU_WAIT_SLEEP_DURATION`: Delay between startup "health checks" (default to `10` seconds).
- `GLUU_MAX_RAM_FRACTION`: Used in conjunction with Docker memory limitations (`docker run -m <mem>`) to identify the fraction of the maximum amount of heap memory you want the JVM to use.
- `GLUU_LDAP_URL`: The LDAP database's IP address or hostname. Default is `localhost:1636`. Multiple URLs can be used using comma-separated values (i.e. `192.168.100.1:1636,192.168.100.2:1636`).
- `GLUU_SHIB_SOURCE_DIR`: absolute path to directory to copy Shibboleth config from (default is `/opt/shared-shibboleth-idp`)
- `GLUU_SHIB_TARGET_DIR`: absolute path to directory to copy Shibboleth config to (default is `/opt/shibboleth-idp`)
- `GLUU_PERSISTENCE_TYPE`: Persistence backend being used (one of `ldap`, `couchbase`, or `hybrid`; default to `ldap`).
- `GLUU_PERSISTENCE_LDAP_MAPPING`: Specify data that should be saved in LDAP (one of `default`, `user`, `cache`, `site`, or `token`; default to `default`). Note this environment only takes effect when `GLUU_PERSISTENCE_TYPE` is set to `hybrid`.
- `GLUU_LDAP_URL`: Address and port of LDAP server (default to `localhost:1636`); required if `GLUU_PERSISTENCE_TYPE` is set to `ldap` or `hybrid`.
- `GLUU_COUCHBASE_URL`: Address of Couchbase server (default to `localhost`); required if `GLUU_PERSISTENCE_TYPE` is set to `couchbase` or `hybrid`.
- `GLUU_COUCHBASE_USER`: Username of Couchbase server (default to `admin`); required if `GLUU_PERSISTENCE_TYPE` is set to `couchbase` or `hybrid`.
- `GLUU_COUCHBASE_CERT_FILE`: Couchbase root certificate location (default to `/etc/certs/couchbase.crt`); required if `GLUU_PERSISTENCE_TYPE` is set to `couchbase` or `hybrid`.
- `GLUU_COUCHBASE_PASSWORD_FILE`: Path to file contains Couchbase password (default to `/etc/gluu/conf/couchbase_password`); required if `GLUU_PERSISTENCE_TYPE` is set to `couchbase` or `hybrid`.
ENCE_TYPE` is set to `couchbase` or `hybrid`.
- `GLUU_COUCHBASE_CONN_TIMEOUT`: Connect timeout used when a bucket is opened (default to `10000` milliseconds).
- `GLUU_COUCHBASE_CONN_MAX_WAIT`: Maximum time to wait before retrying connection (default to `20000` milliseconds).
- `GLUU_COUCHBASE_SCAN_CONSISTENCY`: Default scan consistency; one of `not_bounded`, `request_plus`, or `statement_plus` (default to `not_bounded`).
- `GLUU_DOCUMENT_STORE_TYPE`: Document store type (one of `LOCAL` or `JCA`; default to `LOCAL`).
- `GLUU_JCA_URL`: URL to remote repository (default to `http://localhost:8080`).
- `GLUU_JCA_SYNC_INTERVAL`: Interval between files sync (default to `300` seconds).


CHANGES.md
# Changelog

Here you can see an overview of changes between each release.

## Version 4.2.0_02

Released on July 29th, 2020.

* Fixed missing `bash` executable for Jython script. Reference: https://github.com/GluuFederation/docker-oxshibboleth/issues/11.

## Version 4.2.0_01

Released on July 18th, 2020.

* Added support for Gluu Server v4.2.
* Upraded to Java 11.
* Upgraded to Python3.

## Version 4.1.1_04

Released on June 5th, 2020.

* Added RClone to interact with Jackrabbit via webdav protocol.
* Added feature to cleanup unused metadata XML TR file.
* Upgraded `oxShibbolethIdp`.

## Version 4.1.1_03

Released on April 29th, 2020.

* Added Jackrabbit/JCR support.

## Version 4.1.1_02

Released on April 2nd, 2020.

* Disabled `shibwatcher` script.

## Version 4.1.1_01

Released on March 24th, 2020.

* Conformed to Gluu Server v4.1.1.

## Version 4.1.0_01

Released on March 5th, 2020.

* Conformed to Gluu Server v4.1.

## Version 4.0.1_06

Released on April 2nd, 2020.

* Disabled `shibwatcher` script.

## Version 4.0.1_05

Released on March 5th, 2020.

* Upgraded `oxshibbolethIdp`.
* Added ENV for customizing Couchbase connection and scan consistency.

## Version 4.0.1_04

Released on January 3rd, 2020.

* Fixed invalid tag in `idp-metadata.xml` template.

## Version 4.0.1_03

Released on December 1st, 2019.

* Upgraded `oxshibbolethIdp` v4.0.1.Final build at 2019-11-30.

## Version 4.0.1_02

Released on November 14th, 2019.

* Upgraded `pygluu-containerlib` to show connection issue with Couchbase explicitly.

## Version 4.0.1_01

Released on November 1st, 2019.

* Upgraded to Gluu Server 4.0.1.

## Version 4.0.0_01

Released on October 22nd, 2019.

* Upgraded to Gluu Server 4.0.

## Version 3.1.6_02

Released on May 10th, 2019.

* Alpine upgraded to v3.9. Ref: https://github.com/GluuFederation/gluu-docker/issues/71.

## Version 3.1.6_01

Released on April 29th, 2019.

* Upgraded to Gluu Server 3.1.6.

## Version 3.1.5_03

Released on May 10th, 2019.

* Alpine upgraded to v3.9. Ref: https://github.com/GluuFederation/gluu-docker/issues/71.

## Version 3.1.5_02

Released on April 9th, 2019.

* Added license info on container startup.
* Disabled `sendServerVersion` config of Jetty server.

## Version 3.1.5_01

Released on March 23rd, 2019.

* Upgraded to Gluu Server 3.1.5.

## Version 3.1.4_02

Released on April 4th, 2019.

* Added license info during container run.

## Version 3.1.4_01

Released on November 12th, 2018.

* Upgraded to Gluu Server 3.1.4.

## Version 3.1.3_06

Released on September 18th, 2018.

* Changed base image to use Alpine 3.8.1.

## Version 3.1.3_05

Released on September 12th, 2018.

* Added feature to connect to secure Consul (HTTPS).

## Version 3.1.3_04

Released on August 31st, 2018.

* Added Tini to handle signal forwarding and reaping zombie processes.

## Version 3.1.3_03

Released on July 31st, 2018.

* Fixed template for LDAP password.
* Fixed missing files pulled from shared Shibboleth directory.

## Version 3.1.3_02

Released on July 20th, 2018.

* Added wrapper to manage config via Consul KV or Kubernetes configmap.

## Version 3.1.3_01

Released on June 6th, 2018.

* Upgraded to Gluu Server 3.1.3.

## Version 3.1.2_01

Released on June 6th, 2018.

* Upgraded to Gluu Server 3.1.2.


scripts/wait.py
import logging
import logging.config
import os
import sys

from pygluu.containerlib import get_manager
from pygluu.containerlib import wait_for
from pygluu.containerlib import PERSISTENCE_LDAP_MAPPINGS
from pygluu.containerlib import PERSISTENCE_TYPES

from settings import LOGGING_CONFIG

logging.config.dictConfig(LOGGING_CONFIG)
logger = logging.getLogger("wait")


def main():
    persistence_type = os.environ.get("GLUU_PERSISTENCE_TYPE", "ldap")
    ldap_mapping = os.environ.get("GLUU_PERSISTENCE_LDAP_MAPPING", "default")

    if persistence_type not in PERSISTENCE_TYPES:
        logger.error(
            "Unsupported GLUU_PERSISTENCE_TYPE value; "
            "please choose one of {}".format(", ".join(PERSISTENCE_TYPES))
        )
        sys.exit(1)

    if persistence_type == "hybrid" and ldap_mapping not in PERSISTENCE_LDAP_MAPPINGS:
        logger.error(
            "Unsupported GLUU_PERSISTENCE_LDAP_MAPPING value; "
            "please choose one of {}".format(", ".join(PERSISTENCE_LDAP_MAPPINGS))
        )
        sys.exit(1)

    manager = get_manager()
    deps = ["config", "secret"]

    if persistence_type == "hybrid":
        deps += ["ldap", "couchbase"]
    else:
        deps.append(persistence_type)

    # deps.append("oxtrust")
    wait_for(manager, deps)


if __name__ == "__main__":
    main()


scripts/jca_sync.py
import contextlib
import glob
import logging.config
import os
import time

from pygluu.containerlib.document import RClone

from settings import LOGGING_CONFIG

ROOT_DIR = "/repository/default"
SYNC_DIR = "/opt/shibboleth-idp"

logging.config.dictConfig(LOGGING_CONFIG)
logger = logging.getLogger("webdav")


def sync_from_webdav(url, username, password):
    rclone = RClone(url, username, password)
    rclone.configure()

    logger.info(f"Sync files with remote directory {url}{ROOT_DIR}{SYNC_DIR}")
    rclone.copy_from(SYNC_DIR, SYNC_DIR)


def prune_local_tr(url, username, password):
    def remote_tr_files(files):
        for f in files:
            f = f.strip().split(" ")[1]
            if f.endswith("-sp-metadata.xml"):
                yield f

    rclone = RClone(url, username, password)
    rclone.configure()

    logger.info(f"Removing obsolete local TR files (if any)")
    out = rclone.ls("/opt/shibboleth-idp/metadata")
    files = tuple(remote_tr_files(out.decode().splitlines()))

    for file_ in glob.iglob("/opt/shibboleth-idp/metadata/*-sp-metadata.xml"):
        basename = os.path.basename(file_)
        if basename in files:
            continue

        with contextlib.suppress(FileNotFoundError):
            os.unlink(file_)


def get_sync_interval():
    default = 5 * 60  # 5 minutes

    try:
        interval = int(os.environ.get("GLUU_JCA_SYNC_INTERVAL", default))
    except ValueError:
        interval = default
    return interval


def main():
    store_type = os.environ.get("GLUU_DOCUMENT_STORE_TYPE", "LOCAL")
    if store_type != "JCA":
        logger.warning(f"Using {store_type} document store; sync is disabled ...")
        return

    url = os.environ.get("GLUU_JCA_URL", "http://localhost:8080")
    username = os.environ.get("GLUU_JCA_USERNAME", "admin")
    password = "admin"

    password_file = os.environ.get("GLUU_JCA_PASSWORD_FILE", "/etc/gluu/conf/jca_password")
    if os.path.isfile(password_file):
        with open(password_file) as f:
            password = f.read().strip()

    sync_interval = get_sync_interval()
    try:
        while True:
            sync_from_webdav(url, username, password)
            prune_local_tr(url, username, password)
            time.sleep(sync_interval)
    except KeyboardInterrupt:
        logger.warning("Canceled by user; exiting ...")


if __name__ == "__main__":
    main()


scripts/entrypoint.py
import glob
import os
import re

from pygluu.containerlib import get_manager
from pygluu.containerlib.persistence import render_hybrid_properties
from pygluu.containerlib.persistence import render_couchbase_properties
# from pygluu.containerlib.persistence import sync_couchbase_cert
from pygluu.containerlib.persistence import sync_couchbase_truststore
from pygluu.containerlib.persistence import render_salt
from pygluu.containerlib.persistence import render_gluu_properties
from pygluu.containerlib.persistence import render_ldap_properties
from pygluu.containerlib.persistence import sync_ldap_truststore
from pygluu.containerlib.persistence.couchbase import get_couchbase_mappings
from pygluu.containerlib.persistence.couchbase import get_couchbase_user
from pygluu.containerlib.persistence.couchbase import get_couchbase_password
from pygluu.containerlib.persistence.couchbase import CouchbaseClient
from pygluu.containerlib.utils import decode_text
from pygluu.containerlib.utils import exec_cmd
from pygluu.containerlib.utils import safe_render
from pygluu.containerlib.utils import cert_to_truststore
from pygluu.containerlib.utils import get_server_certificate

GLUU_LDAP_URL = os.environ.get("GLUU_LDAP_URL", "localhost:1636")
GLUU_COUCHBASE_URL = os.environ.get("GLUU_COUCHBASE_URL", "localhost")


def render_idp3_templates(manager):
    ldap_hostname, ldaps_port = GLUU_LDAP_URL.split(":")

    persistence_type = os.environ.get("GLUU_PERSISTENCE_TYPE", "ldap")
    ldap_mapping = os.environ.get("GLUU_PERSISTENCE_LDAP_MAPPING", "default")

    idp_resolver_filter = "(|(uid=$requestContext.principalName)(mail=$requestContext.principalName))"

    if all([persistence_type in ("couchbase", "hybrid"),
            "user" in get_couchbase_mappings(persistence_type, ldap_mapping)]):
        idp_resolver_filter = "(&(|(lower(uid)=$requestContext.principalName)(mail=$requestContext.principalName))(objectClass=gluuPerson))"

    ctx = {
        "hostname": manager.config.get("hostname"),
        "shibJksPass": manager.secret.get("shibJksPass"),
        "certFolder": "/etc/certs",
        "ldap_hostname": ldap_hostname,
        "ldaps_port": ldaps_port,
        "ldap_binddn": manager.config.get("ldap_binddn"),
        "ldapPass": decode_text(manager.secret.get("encoded_ox_ldap_pw"), manager.secret.get("encoded_salt")),
        "idp3SigningCertificateText": load_cert_text("/etc/certs/idp-signing.crt"),
        "idp3EncryptionCertificateText": load_cert_text("/etc/certs/idp-encryption.crt"),
        "orgName": manager.config.get("orgName"),
        "ldapCertFn": "/etc/certs/opendj.crt",
        "couchbase_hostname": GLUU_COUCHBASE_URL,
        "couchbaseShibUserPassword": manager.secret.get("couchbase_shib_user_password"),
        "idp_attribute_resolver_ldap.search_filter": idp_resolver_filter,
    }

    for file_path in glob.glob("/app/templates/idp3/*.properties"):
        with open(file_path) as fr:
            rendered_content = safe_render(fr.read(), ctx)
            fn = os.path.basename(file_path)
            with open("/opt/shibboleth-idp/conf/{}".format(fn), 'w') as fw:
                fw.write(rendered_content)

    file_path = "/app/templates/idp3/idp-metadata.xml"
    with open(file_path) as fr:
        rendered_content = safe_render(fr.read(), ctx)
        fn = os.path.basename(file_path)
        with open("/opt/shibboleth-idp/metadata/{}".format(fn), 'w') as fw:
            fw.write(rendered_content)


def load_cert_text(path):
    with open(path) as f:
        cert = f.read()
        return cert.replace('-----BEGIN CERTIFICATE-----', '').replace('-----END CERTIFICATE-----', '').strip()


def generate_idp3_sealer(manager):
    cmd = " ".join([
        "java",
        "-classpath '/opt/gluu/jetty/idp/webapps/idp/WEB-INF/lib/*'",
        "net.shibboleth.utilities.java.support.security.BasicKeystoreKeyStrategyTool",
        "--storefile /opt/shibboleth-idp/credentials/sealer.jks",
        "--versionfile /opt/shibboleth-idp/credentials/sealer.kver",
        "--alias secret",
        "--storepass {}".format(manager.secret.get("shibJksPass")),
    ])
    return exec_cmd(cmd)


def sync_sealer(manager):
    jks_fn = "/opt/shibboleth-idp/credentials/sealer.jks"
    kver_fn = "/opt/shibboleth-idp/credentials/sealer.kver"

    if all([os.path.isfile(jks_fn), os.path.isfile(kver_fn)]):
        return

    # files are missing, get them from secrets (if any) or generate new ones
    try:
        manager.secret.to_file("sealer_jks_base64", jks_fn, decode=True, binary_mode=True)
        manager.secret.to_file("sealer_kver_base64", kver_fn, decode=True, binary_mode=True)
    except TypeError:
        generate_idp3_sealer(manager)
        manager.secret.from_file("sealer_jks_base64", jks_fn, encode=True, binary_mode=True)
        manager.secret.from_file("sealer_kver_base64", kver_fn, encode=True, binary_mode=True)


def modify_jetty_xml():
    fn = "/opt/jetty/etc/jetty.xml"
    with open(fn) as f:
        txt = f.read()

    # disable contexts
    updates = re.sub(
        r'<New id="DefaultHandler" class="org.eclipse.jetty.server.handler.DefaultHandler"/>',
        r'<New id="DefaultHandler" class="org.eclipse.jetty.server.handler.DefaultHandler">\n\t\t\t\t <Set name="showContexts">false</Set>\n\t\t\t </New>',
        txt,
        flags=re.DOTALL | re.M,
    )

    # disable Jetty version info
    updates = re.sub(
        r'(<Set name="sendServerVersion"><Property name="jetty.httpConfig.sendServerVersion" deprecated="jetty.send.server.version" default=")true(" /></Set>)',
        r'\1false\2',
        updates,
        flags=re.DOTALL | re.M,
    )

    with open(fn, "w") as f:
        f.write(updates)


def modify_webdefault_xml():
    fn = "/opt/jetty/etc/webdefault.xml"
    with open(fn) as f:
        txt = f.read()

    # disable dirAllowed
    updates = re.sub(
        r'(<param-name>dirAllowed</param-name>)(\s*)(<param-value>)true(</param-value>)',
        r'\1\2\3false\4',
        txt,
        flags=re.DOTALL | re.M,
    )

    with open(fn, "w") as f:
        f.write(updates)


def saml_couchbase_settings():
    # Add couchbase bean to global.xml
    global_xml_fn = "/opt/shibboleth-idp/conf/global.xml"
    with open(global_xml_fn) as f:
        global_xml = f.read()

    with open("/app/static/couchbase_bean.xml") as f:
        bean_xml = f.read()

    with open(global_xml_fn, "w") as f:
        global_xml = global_xml.replace("</beans>", bean_xml + "\n\n</beans>")
        f.write(global_xml)

    # Add datasource.properties to idp.properties
    idp_properties_fn = "/opt/shibboleth-idp/conf/idp.properties"
    with open(idp_properties_fn) as f:
        idp3_properties = f.readlines()

    for i, l in enumerate(idp3_properties):
        if l.strip().startswith('idp.additionalProperties'):
            idp3_properties[i] = l.strip() + ', /conf/datasource.properties\n'

    with open(idp_properties_fn, "w") as f:
        new_idp3_props = ''.join(idp3_properties)
        f.write(new_idp3_props)


def create_couchbase_shib_user(manager):
    hostname = GLUU_COUCHBASE_URL
    user = get_couchbase_user(manager)
    password = get_couchbase_password(manager)

    cb_client = CouchbaseClient(hostname, user, password)
    cb_client.create_user(
        'couchbaseShibUser',
        manager.secret.get("couchbase_shib_user_password"),
        'Shibboleth IDP',
        'query_select[*]',
    )


def main():
    persistence_type = os.environ.get("GLUU_PERSISTENCE_TYPE", "ldap")
    ldap_mapping = os.environ.get("GLUU_PERSISTENCE_LDAP_MAPPING", "default")
    manager = get_manager()

    if not os.path.isfile("/etc/certs/idp-signing.crt"):
        manager.secret.to_file("idp3SigningCertificateText", "/etc/certs/idp-signing.crt")

    if not os.path.isfile("/etc/certs/idp-signing.key"):
        manager.secret.to_file("idp3SigningKeyText", "/etc/certs/idp-signing.key")

    if not os.path.isfile("/etc/certs/idp-encryption.crt"):
        manager.secret.to_file("idp3EncryptionCertificateText", "/etc/certs/idp-encryption.crt")

    if not os.path.isfile("/etc/certs/idp-encryption.key"):
        manager.secret.to_file("idp3EncryptionKeyText", "/etc/certs/idp-encryption.key")

    manager.secret.to_file("shibIDP_jks_base64", "/etc/certs/shibIDP.jks",
                           decode=True, binary_mode=True)

    sync_sealer(manager)

    render_idp3_templates(manager)
    render_salt(manager, "/app/templates/salt.tmpl", "/etc/gluu/conf/salt")
    render_gluu_properties("/app/templates/gluu.properties.tmpl", "/etc/gluu/conf/gluu.properties")

    if persistence_type in ("ldap", "hybrid"):
        render_ldap_properties(
            manager,
            "/app/templates/gluu-ldap.properties.tmpl",
            "/etc/gluu/conf/gluu-ldap.properties",
        )
        manager.secret.to_file("ldap_ssl_cert", "/etc/certs/opendj.crt", decode=True)
        sync_ldap_truststore(manager)

    if persistence_type in ("couchbase", "hybrid"):
        render_couchbase_properties(
            manager,
            "/app/templates/gluu-couchbase.properties.tmpl",
            "/etc/gluu/conf/gluu-couchbase.properties",
        )
        # sync_couchbase_cert(manager)
        sync_couchbase_truststore(manager)
        create_couchbase_shib_user(manager)

        if "user" in get_couchbase_mappings(persistence_type, ldap_mapping):
            saml_couchbase_settings()

    if persistence_type == "hybrid":
        render_hybrid_properties("/etc/gluu/conf/gluu-hybrid.properties")

    if not os.path.isfile("/etc/certs/gluu_https.crt"):
        get_server_certificate(manager.config.get("hostname"), 443, "/etc/certs/gluu_https.crt")

    cert_to_truststore(
        "gluu_https",
        "/etc/certs/gluu_https.crt",
        "/usr/lib/jvm/default-jvm/jre/lib/security/cacerts",
        "changeit",
    )

    modify_jetty_xml()
    modify_webdefault_xml()


if __name__ == "__main__":
    main()


scripts/wait_for.py
import argparse
import base64
import json
import logging
import os
import random
import sys
import time

import ldap3
import pyDes

from cbm import CBM
from gluulib import get_manager

logger = logging.getLogger("wait_for")
logger.setLevel(logging.INFO)
ch = logging.StreamHandler()
fmt = logging.Formatter('%(levelname)s - %(asctime)s - %(message)s')
ch.setFormatter(fmt)
logger.addHandler(ch)


def decode_password(manager, password_key, salt_key):
    encoded_password = manager.secret.get(password_key)
    encoded_salt = manager.secret.get(salt_key)

    cipher = pyDes.triple_des(
        b"{}".format(encoded_salt),
        pyDes.ECB,
        padmode=pyDes.PAD_PKCS5
    )
    encrypted_text = b"{}".format(base64.b64decode(encoded_password))
    return cipher.decrypt(encrypted_text)


def wait_for_config(manager, max_wait_time, sleep_duration):
    for i in range(0, max_wait_time, sleep_duration):
        try:
            reason = "config 'hostname' is not available"
            if manager.config.get("hostname"):
                logger.info("Config backend is ready.")
                return
        except Exception as exc:
            reason = exc

        logger.warn("Config backend is not ready; reason={}; "
                    "retrying in {} seconds.".format(reason, sleep_duration))
        time.sleep(sleep_duration)

    logger.error("Config backend is not ready after {} seconds.".format(max_wait_time))
    sys.exit(1)


def wait_for_secret(manager, max_wait_time, sleep_duration):
    for i in range(0, max_wait_time, sleep_duration):
        try:
            reason = "secret 'ssl_cert' is not available"
            if manager.secret.get("ssl_cert"):
                logger.info("Secret backend is ready.")
                return
        except Exception as exc:
            reason = exc

        logger.warn("Secret backend is not ready; reason={}; "
                    "retrying in {} seconds.".format(reason, sleep_duration))
        time.sleep(sleep_duration)

    logger.error("Secret backend is not ready after {} seconds.".format(max_wait_time))
    sys.exit(1)


def wait_for_ldap(manager, max_wait_time, sleep_duration):
    GLUU_LDAP_URL = os.environ.get("GLUU_LDAP_URL", "localhost:1636")
    GLUU_PERSISTENCE_LDAP_MAPPING = os.environ.get("GLUU_PERSISTENCE_LDAP_MAPPING", "default")

    ldap_bind_dn = manager.config.get("ldap_binddn")
    ldap_password = decode_password(manager, "encoded_ox_ldap_pw", "encoded_salt")

    ldap_host = GLUU_LDAP_URL.split(":")[0]
    ldap_port = int(GLUU_LDAP_URL.split(":")[1])

    ldap_server = ldap3.Server(
        ldap_host,
        ldap_port,
        use_ssl=True
    )

    # check the entries few times, to ensure OpenDJ is running after importing
    # initial data; this may not required for OpenLDAP
    successive_entries_check = 0

    search_base_mapping = {
        "default": "o=gluu",
        "user": "o=gluu",
        "site": "o=site",
        "cache": "o=gluu",
        "statistic": "o=metric",
    }
    search_base = search_base_mapping[GLUU_PERSISTENCE_LDAP_MAPPING]

    for i in range(0, max_wait_time, sleep_duration):
        try:
            with ldap3.Connection(
                    ldap_server,
                    ldap_bind_dn,
                    ldap_password) as ldap_connection:

                ldap_connection.search(
                    search_base=search_base,
                    search_filter="(objectClass=*)",
                    search_scope=ldap3.SUBTREE,
                    attributes=['objectClass'],
                    size_limit=1,
                )

                if ldap_connection.entries:
                    successive_entries_check += 1

                if successive_entries_check >= 3:
                    logger.info("LDAP is ready")
                    return
                reason = "LDAP is not initialized yet"
        except Exception as exc:
            reason = exc

        logger.warn("LDAP backend is not ready; reason={}; "
                    "retrying in {} seconds.".format(reason, sleep_duration))
        time.sleep(sleep_duration)

    logger.error("LDAP not ready, after " + str(max_wait_time) + " seconds.")
    sys.exit(1)


def check_couchbase_document(cbm):
    persistence_type = os.environ.get("GLUU_PERSISTENCE_TYPE", "ldap")
    ldap_mapping = os.environ.get("GLUU_PERSISTENCE_LDAP_MAPPING", "default")
    checked = True
    error = ""
    bucket = "gluu"

    if persistence_type == "hybrid":
        req = cbm.get_buckets()
        if not req.ok:
            checked = False
            error = json.loads(req.text)["errors"][0]["msg"]
            return checked, error

        bucket = random.choice([
            _bucket["name"] for _bucket in req.json()
            if _bucket["name"] != ldap_mapping
        ])

    query = "SELECT COUNT(*) FROM `{}`".format(bucket)
    req = cbm.exec_query(query)
    if not req.ok:
        checked = False
        error = json.loads(req.text)["errors"][0]["msg"]
    return checked, error


def wait_for_couchbase(manager, max_wait_time, sleep_duration):
    hostname = os.environ.get("GLUU_COUCHBASE_URL", "localhost")
    user = manager.config.get("couchbase_server_user")
    cbm = CBM(hostname, user, decode_password(
        manager, "encoded_couchbase_server_pw", "encoded_salt",
    ))

    for i in range(0, max_wait_time, sleep_duration):
        try:
            checked, error = check_couchbase_document(cbm)
            if checked:
                logger.info("Couchbase is ready")
                return
            reason = error
        except Exception as exc:
            reason = exc

        logger.warn("Couchbase backend is not ready; reason={}; "
                    "retrying in {} seconds.".format(reason, sleep_duration))
        time.sleep(sleep_duration)

    logger.error("Couchbase backend is not ready after {} seconds.".format(max_wait_time))
    sys.exit(1)


def wait_for(manager, deps=None):
    deps = deps or []

    try:
        max_wait_time = int(os.environ.get("GLUU_WAIT_MAX_TIME", 300))
    except ValueError:
        max_wait_time = 300

    try:
        sleep_duration = int(os.environ.get("GLUU_WAIT_SLEEP_DURATION", 5))
    except ValueError:
        sleep_duration = 5

    if "config" in deps:
        wait_for_config(manager, max_wait_time, sleep_duration)

    if "secret" in deps:
        wait_for_secret(manager, max_wait_time, sleep_duration)

    if "ldap" in deps:
        wait_for_ldap(manager, max_wait_time, sleep_duration)

    if "couchbase" in deps:
        wait_for_couchbase(manager, max_wait_time, sleep_duration)


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--deps", help="Comma-separated dependencies to wait for.")
    args = parser.parse_args()

    deps = set(filter(
        None,
        [dep.strip() for dep in args.deps.split(",") if dep]
    ))

    manager = get_manager()
    wait_for(manager, deps)
